### ISSUE TITLE
Possible optimization? Automatically downgrade integer types.

### ISSUE THREAD

**Issue by danilaml** [AliveToolkit/alive2#729]
While checking a simple opt on [alive2](https://alive2.llvm.org/ce/z/ugbo6U) (thanks for the tool, btw, it's really cool), I've noticed that it usually times out with `i32` integer types but works fine if I replace all of them with smaller type, i.e. `i16`.
And for the program in question, the specific type doesn't appear to matter.
I was wondering if it's possible to detect such class of programs (i.e. just generic width-agnostic IR instructions, no specific immediates, other than all 1/0 and similar)  and automatically do the verification in the smallest/fastest possible type?

**Comment by regehr** [comment]
there are plenty of cases where this could be done, but in general detecting "bitwidth independence" is not easy. 

just as a random example, it is not trivial to create a smaller-width version of the example we picked for the compiler explorer instance https://alive2.llvm.org/ce/z/XxT-iw

**Comment by danilaml** [comment]
Yeah, that'd fall under the case of "specific constants". Don't know how easy it'd be to define some useful set that is "bitwidth independent" in that way.
